fix(performance): Fix n+1 issue when fetching calendar properties

CarlSchwan opened the pull request:
## Summary

Fix issue where we are requesting for each calendar individually the properties while we could query the properties for each calendar at the same time.

Before (with only 3 calendars):

<img width="1089" height="224" alt="image" src="https://github.com/user-attachments/assets/e427d926-481a-4a38-94b8-5d2a209f1ec6" />

After:

<img width="730" height="245" alt="image" src="https://github.com/user-attachments/assets/5546ee59-7d02-4c84-b4f9-47ca81432c75" />

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

Commit:
perf(caldav): Fetch additional properties for multiple calendars in one query

Instead of doing it for each calendar individually.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

Commit:
perf(caldav): Add multi users support in findPropertiesByPaths

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

st3iny reviewed (approved):
Awesome! It's even worse on my local dev/testing instance where there are 40 individual queries for `oc_properties` on the PROPFIND :see_no_evil:

st3iny commented:
Code checkers failure is not related. Some translations seem to be broken on master ...

```
Run php ./build/translation-checker.php
  
ERROR: There were 1 errors:
/home/runner/actions-runner/_work/server/server/build/../apps/files/l10n/vi.json
  Contains a RTL limited characters in the translations. Offending strings:
    "Toggle grid view" : "‎Chuyển đổi dạng xem lưới‎",
```

Commit:
perf(caldav): Cache calendars in CustomPropertiesBackend

We already do that for files, we are now also doing for calendars.
With relatively small amount of calendars, I managed to reduce the
number of DB requests by 35% and from 23 DB requests touching the
oc_properties table to only 3.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

salmart-dev commented on apps/dav/lib/Db/PropertyMapper.php:
```diff
@@ -39,17 +40,36 @@ public function findPropertyByPathAndName(string $userId, string $path, string $
+			->from(self::TABLE_NAME);
+
+		foreach ($calendars as $user => $paths) {
+			$selectQb->andWhere(
```
Wouldn't this query always return no rows? The query built would be something like `... WHERE userId = ? AND propertypath IN ? AND userId = ? AND ...`, I think this would never match?

CarlSchwan replied:
good catch fixed with 3fd0b11580a5cc39fe66923d816de9a5d46f238d

salmart-dev commented on apps/dav/lib/DAV/CustomPropertiesBackend.php:
```diff
@@ -376,6 +394,62 @@ private function cacheDirectory(string $path, Directory $node): void {
+		}
+		$paths = array_unique($paths);
+
+		$propsByPath = [];
+		foreach ($paths as $path) {
+			$propsByPath[$path] = [];
+		}
```
This could be shortened to:

```suggestion
		$propsByPath = array_fill_keys(array_values($paths), []);
```

salmart-dev commented on apps/dav/lib/DAV/CustomPropertiesBackend.php:
```diff
@@ -376,6 +394,62 @@ private function cacheDirectory(string $path, Directory $node): void {
+		foreach ($users as $nestedPaths) {
+			$paths = array_merge($paths, $nestedPaths);
+		}
+		$paths = array_unique($paths);
```
From what I understand from the code, this will not yield much different results than the previous iteration for the user cache. The only difference is that the paths will be unique, but from what I understand we would load all properties for all the calendars that match this path. Is this intentional?

CarlSchwan replied:
let's try to do that in a follow up :)

Commit:
perf(caldav): Only prefetch published properties

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

icewind1991 commented on apps/dav/lib/CalDAV/CalendarProvider.php:
```diff
@@ -36,9 +36,14 @@ public function getCalendars(string $principalUri, array $calendarUris = []): ar
+			$user = str_replace('principals/users/', '', $calendarInfo['principaluri']);
+			$path = 'calendars/' . $user . '/' . $calendarInfo['uri'];
+
+			$calendarInfo = array_merge($calendarInfo, array_key_exists($path, $additionalProperties) ? $additionalProperties[$path] : []);
```
`$additionalProperties[$path] ?? []` ?

CarlSchwan commented on apps/dav/lib/Db/PropertyMapper.php:
```diff
@@ -39,17 +40,43 @@ public function findPropertyByPathAndName(string $userId, string $path, string $
+			->from(self::TABLE_NAME);
+
+		foreach ($calendars as $user => $paths) {
+			$selectQb->orWhere(
+				$selectQb->expr()->andX(
+					$selectQb->expr()->eq('userid', $selectQb->createNamedParameter($user)),
+					$selectQb->expr()->in('propertypath', $selectQb->createNamedParameter($paths, IQueryBuilder::PARAM_STR_ARRAY)),
+				)
 			);
```
Yeah, it should be fine and still better than doing n-queries

CarlSchwan commented on apps/dav/lib/Db/PropertyMapper.php:
```diff
@@ -39,17 +40,43 @@ public function findPropertyByPathAndName(string $userId, string $path, string $
 	}
 
 	/**
+	 * @param array<string, string[]> $calendars
```
this made psalm unhappy :/
